states.go

```diff
@@ -25,5 +25,7 @@ func (c *Client) ReadStates(ctx context.Context) (command.StatesCommand, error)
 		return nil, fmt.Errorf("error while waiting for device states: %w", err)
 	}
 
+	c.stateChanged = false
```

Comment:
> you only need a ble-device-address and must not establish a "whole" nuki-client.

For the pure monitoring, yes. However, literally the only thing the advertisement bit tells you is that _any_ state has changed. You'll need to call `readState()` to learn which of the states it was and what the new state is - and for that you'll need a full client. I don't think there'll be any script which can work without a client, and if you create it anyway, then why not pass it?

If you just gave a device address to the monitoring method, then the callback could also tell you only the address. You would have to create a new client every time, pass the authentication details again etc. Or you would need your own lookup map to find your pre-created client for that address.

The way I implemented it, you can create the client _once_ in the beginning of your script, set it up with all the details(*) and then start monitoring it. The callback will be triggered with that ready-to-use-client as parameter, so you can immediately establish a connection, call `readState()` and whatever else you want.

I'm all for keeping things flexible, but again, I don't see the need to "save" the creation of a client and would rather make it easy to use.

---
_(*) I have prepared (but not yet pushed) a helper for easier (un-)marshalling from/to JSON and YAML, so you can keep the address and credentials in a config and easily create the client(s) from that._

Reply:
> In my opinion the multi-inform-issue should be an issue for the "StateChangeHandler" and not an issue for the nuki-library.

At least my Nuki Smart Lock 3.0 sends BLE advertisements about once per second. Each of them includes the flag, so as soon as someone e.g. unlocks via app, the callback would be triggered every second. I can't think of any use-case where a script would need repeating notifications, because again, there isn't any state encoded in the bit itself. Therefore I think the library should take care of it.

Reply:
I thought about the Addr vs. Client topic a bit more. Actually the Client struct could be split into two groups of fields:
1) Static information/configuration
    - addr (added by me)
    - privateKey
    - publicKey
    - nukiPublicKey
    - authId
    - pin (potential future extension)
2) State/connection resources:
    - responseTimeout (although that's somewhere middle ground and could go in either section)
    - client
    - gdioCom
    - udioCom

That's obvious when we look at `Close()` - only the second group of fields is reset there. The first group of fields exists even when disconnected, it's just a few bytes in main memory, which is far cheaper to create than something involving the Bluetooth stack.

For a typical scenario with monitoring, the local Bluetooth device will be in scanning state most of the time, and will only establish a connection for a short time when either the lock signals that its state has changed or when some action (e.g. unlock) has been requested by the user. For this scenario, the struct with the static information would be created once when the script starts, and the connection struct would be created and destroyed only when needed.

Looking at `go-ble`, they have a similar thing: You have the `Addr` as configuration and receive a `Client` from `Dial()`. Following the same approach, this library could have the above field groups as `Config` and `Client` structs. `Client` would have a reference to the underlying `Config`. Most existing methods would stay in `Client`, biggest exception being `EstablishConnection()` which would have to be called on `Config` and would create a new `Client`.

Not sure if you'd be ready to break backwards-compatibility that much. It would separate concerns and might make it easier to justify that a `Config` (but not a `Client`) should be used to monitor devices. Maybe I'll just go ahead and prepare a PR to see what the result looks like.

Reply:
> Maybe I'll just go ahead and prepare a PR to see what the result looks like.

Check out https://github.com/rovo89/go-nuki/commits/split_client. Maybe not completely done yet, but should give a rough idea into which direction it could go. Btw, `Config` might be a bad name because it could be confused with `ReadConfig()`. But I couldn't think of anything better, only `Device`, but that's also used be go-ble for the local Bluetooth devices...

Reply:
Yeah maybe you are right: there is no scenario which the user wants to only inform about changes without knowing *what* changed. Therefore you can leave it how it currently is.

But i will leave some code comments for the current approach.